platform/Span.h

```diff
@@ -0,0 +1,578 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018-2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_PLATFORM_SPAN_H_
+#define MBED_PLATFORM_SPAN_H_
+
+#include <algorithm>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "platform/mbed_assert.h"
+
+namespace mbed {
+
+/**
+ * Special value for the Size parameter of Span.
+ * If the type use this value then the size of the array is stored in the object
+ * at runtime.
+ */
+#define SPAN_DYNAMIC_SIZE -1
```

Comment:
Just for more naming consistency with C++, should the the template parameter and this constant be "extent" rather than "size"?

Feels a bit wonky, but I guess it's intended to allow you to separately query `extent` from `size` in template nonsense. (`if (span.extent == std::dynamic_extent) ...`)

Reply:
I will fix the name of the dynamic tag however it is not meant to be queried at runtime as this information is not really useful. 

Span with fixed extent are mostly useful for type(def) that requires a fixed size otherwise the dynamic extent should be used. Also, the conversion rules offer implicit conversion from `Span<T, Size>` to `Span<T>` so users don't have to worry about conversions.